p2p/p2p-core/src/lib.rs

```diff
@@ -206,55 +236,48 @@ pub trait CoreSyncSvc:
         CoreSyncDataRequest,
         Response = CoreSyncDataResponse,
         Error = tower::BoxError,
-        Future = Pin<
-            Box<
-                dyn Future<Output = Result<CoreSyncDataResponse, tower::BoxError>> + Send + 'static,
-            >,
-        >,
+        Future = Self::Future2,
     > + Send
     + 'static
 {
+    // This allows us to put more restrictive bounds on the future without defining the future here
+    // explicitly.
+    type Future2: Future<Output = Result<Self::Response, Self::Error>> + Send + 'static;
 }
 
-impl<T> CoreSyncSvc for T where
+impl<T> CoreSyncSvc for T
+where
     T: tower::Service<
             CoreSyncDataRequest,
             Response = CoreSyncDataResponse,
             Error = tower::BoxError,
-            Future = Pin<
-                Box<
-                    dyn Future<Output = Result<CoreSyncDataResponse, tower::BoxError>>
-                        + Send
-                        + 'static,
-                >,
-            >,
         > + Send
-        + 'static
+        + 'static,
+    T::Future: Future<Output = Result<Self::Response, Self::Error>> + Send + 'static,
 {
+    type Future2 = T::Future;
```

Comment:
This change looks complicated but is just changing the `Future` type to be generic instead of `Pin<Box<>>`.